src/devices/Mcp25xxx/Mcp25xxx.cs

```diff
@@ -0,0 +1,404 @@
+
+using System;
+using System.Device.Gpio;
+using System.Device.Spi;
+using Iot.Device.Mcp25xxx.Register;
+
+namespace Iot.Device.Mcp25xxx
+{
+    /// <summary>
+    /// A general purpose driver for the Microchip MCP25 CAN controller device family.
+    /// </summary>
+    public abstract class Mcp25xxx : IDisposable
+    {
+        internal IGpioController _masterGpioController;
+        private SpiDevice _spiDevice;
+        private readonly int _reset;
+        private readonly int _tx0rts;
+        private readonly int _tx1rts;
+        private readonly int _tx2rts;
+        private readonly int _interrupt;
+        private readonly int _rx0bf;
+        private readonly int _rx1bf;
+        private readonly int _clkout;
+
+        /// <summary>
+        /// A general purpose driver for the Microchip MCP25 CAN controller device family.
+        /// </summary>
+        /// <param name="spiDevice">The SPI device used for communication.</param>
+        /// <param name="reset">The output pin number that is connected to Reset.</param>
+        /// <param name="tx0rts">The output pin number that is connected to Tx0RTS.</param>
+        /// <param name="tx1rts">The output pin number that is connected to Tx1RTS.</param>
+        /// <param name="tx2rts">The output pin number that is connected to Tx2RTS.</param>
+        /// <param name="interrupt">The input pin number that is connected to INT.</param>
+        /// <param name="rx0bf">The input pin number that is connected to Rx0BF.</param>
+        /// <param name="rx1bf">The input pin number that is connected to Rx1BF.</param>
+        /// <param name="clkout">The input pin number that is connected to CLKOUT.</param>
+        /// <param name="masterGpioController">
+        /// The GPIO controller for defined external pins. If not specified, the default controller will be used.
+        /// </param>
+        public Mcp25xxx(
+            SpiDevice spiDevice,
+            int reset = -1,
+            int tx0rts = -1,
+            int tx1rts = -1,
+            int tx2rts = -1,
+            int interrupt = -1,
+            int rx0bf = -1,
+            int rx1bf = -1,
+            int clkout = -1,
+            IGpioController masterGpioController = null)
+        {
+            _spiDevice = spiDevice;
+
+            _reset = reset;
+            _tx0rts = tx0rts;
+            _tx1rts = tx1rts;
+            _tx2rts = tx2rts;
+            _interrupt = interrupt;
+            _rx0bf = rx0bf;
+            _rx1bf = rx1bf;
+            _clkout = clkout;
+
+            // Only need master controller if there are external pins provided.
+            if (_reset != -1 ||
+                _tx0rts != -1 ||
+                _tx1rts != -1 ||
+                _tx2rts != -1 ||
+                _interrupt != -1 ||
+                _rx0bf != -1 ||
+                _rx1bf != -1 ||
+                _clkout != -1)
+            {
+                _masterGpioController = masterGpioController ?? new GpioController();
```

Comment:
should you throw if not enough pins provided?

Reply:
Do you mean checking the pin count of controller first?  If so, then yes.  In most cases a user probably wouldn't use all pins, but a dev board (RPi) or expander (Mcp23) should provide enough pins if needed.  BTW, this is when #215 would come in handy as it wouldn't force to use one specific GpioController.

Reply:
@shaggygi I was going to comment about using different controllers for different pins but realized you can already do that by implementing your own GpioController and forwarding calls to multiple different controllers (it probably should be fairly easy to create such abstraction in generic way) - you'd need to remap pin numbers though

Reply:
Oh, so you mean have a special IGpioController implementation that could have multiple IGpioControllers it controls?  For example, have a controller that can use 2 Mcp23xxx to control 32 I/O (16x2).

```csharp
public class MyGpioController(IGpioController controller1, IGpioController controller2)

// Setup Mcp23xxx  objects...

var myController = MyGpioController(mcp23xxx1, mcp23xxx2);

// Writing to pin 20, the controller would know it needs to write to respective pin on 2nd mcp23xxx.
myController.Write(20, PinValue.Low);
```

If so, that's a pretty clever idea 👍 

Related to checking pin count... we have to wait until [this issue](https://github.com/dotnet/iot/issues/260#issuecomment-468095697) is complete.  I don't want that issue to hold this PR up.  We can update later if needed.

Reply:
Yes - I'd eventually want to have something like GpioControllerMapper or something like that where you can specify a list of controllers and what addresses they should be mapped to (or rather what number to add to the original pin numbers). This way you can have say Raspberry Pi pins the same as original, Mcp23xxx pins on 100-115, 200-215 (add 100, 200 etc).

I think for sanity might be worth to add some API telling what is the maximum pin number you can find on the given gpio controller

Reply:
> I think for sanity might be worth to add some API telling what is the maximum pin number you can find on the given gpio controller

I thought the PinCount property provided this?  RaspberryPi3Driver has the following...

`protected internal override int PinCount => 28;`